src/shard.c

```diff
@@ -16,78 +16,113 @@
 #include "shard.h"
 #include "state.h"
 
-void shard_init(shard_t *shard, uint8_t shard_id, uint8_t num_shards,
-		uint8_t sub_id, uint8_t num_subshards, const cycle_t *cycle,
+static uint32_t shard_roll_to_valid(shard_t *s)
+{
+	if (s->current - 1 < zsend.max_index) {
+		return s->current;
+	}
+	return shard_get_next_ip(s);
+}
+
+void shard_init(shard_t *shard, uint8_t shard_idx, uint8_t num_shards,
+		uint8_t thread_idx, uint8_t num_threads,
+		uint32_t max_total_targets, const cycle_t *cycle,
 		shard_complete_cb cb, void *arg)
 {
-	// Start out by figuring out the multiplication factor for this shard.
-	// With one shard, this would just be the generator, but with n shards,
-	// f = g^n.
+	// Start out by figuring out how many shards we have. A single shard of
+	// ZMap (set with --shards=N, --shard=n) may have several subshards, if
+	// ZMap is being ran multithreaded (set with --sender-threads=T).
+	//
+	// Total number of subshards is S = N*T. Subshard ID's range from [0,
+	// N*T).
+	assert(num_shards > 0);
+	assert(num_threads > 0);
+	assert(shard_idx < num_shards);
+	assert(thread_idx < num_threads);
+	uint32_t num_subshards = (uint32_t)num_shards * (uint32_t)num_threads;
```

Comment:
This looks like a potential overflow.

Reply:
These are both upcasted uint8_t's, so no overflow.